### PR TITLE
sfcodes harmonization for xls20 and hooks

### PR DESCRIPTION
**Pull request by RichardAH**
## High Level Overview of Change
Hooks and XLS20 both require new serialised field codes in rippled. These codes are currently in conflict.
Hooks has previously had a conflict with Tickets which has [caused considerable confusion.](https://twitter.com/intelliot/status/1440172613955842049)

Our ecosystem does not currently have a system for assigning codes to new amendments. This needs to be addressed.

Allowing conflicts to continue is detrimental to third party tools, applications and wallets which need to serialise and deserialise rippled's data in order to function. These include but are not limited to the ripple-binary-codec and binary visualizer tool.

This PR is an attempt to resolve the current sfcode conflicts between XLS20 and Hooks. Other pending amendments are not included.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

**Comment by WietseWind** [comment]
@nbougalis Hi Nik, can you + team take a good look at this and consider merging? I think we can all agree Hooks are something we do not want to rush, but is definitely something serious enough to reserve codes. We want to prevent conflicts like happened with Tickets.

**Comment by nbougalis** [comment]
@scottschurr and I will review with a goal of getting this into a 1.9.0-b1; we will rebase #4101 on top of this.

**Comment by WietseWind** [comment]
> @scottschurr and I will review with a goal of getting this into a 1.9.0-b1; we will rebase #4101 on top of this.

Appreciate it. A lot :)

**Comment by RichardAH** [comment]
See above referenced commit for just how annoying it can be to have your sfcodes clobbered. Reiterating that we need a better system going forward if multiple people are going to work simultaneously on amendments that change or add sfFields.

**Comment by scottschurr** [comment]
@RichardAH, I respect the pain you have felt as new serialization types are added to the system while you're maintaining code.  I think reserving the codes that Hooks needs is a good idea.

I have two concerns with this commit as it currently stands:
1. Visibility of field usage, and
2. Order of operations.

**Visibility of Field Usage**

Hooks adds quite a number of serialized fields, and that's fine.  What's not so good is that we don't see these serialized fields being used anywhere in the code base as it stands.  I get why that is.  But it's not uncommon for me to look in the SFields.cpp file and wonder, "Hmmm, who is using that field?"  With the newly added Hooks fields if I look in the code base my answer will be nobody.  But that's not actually the case.  They are used, by Hooks, but I can't see that usage.

Additionally I am of the opinion that if something is not tested then it often doesn't work as expected.  This commit adds Hooks fields, but does not exercise them in any way.  That's probably okay?  But it leaves me uneasy.

I'd like to suggest that, in addition to adding the Hooks-specific SFields, you add a Hooks stub unit test file.  It doesn't need to do anything other than:

1. Include a paragraph comment at the top saying what it represents: These are tests that exercise facilities that are reserved for when Hooks is merged in the future.

2. Do some minimal exercising of all the newly added fields.  That way if someone looks for where the fields are used they will be led to this hooks-related test file.

This solves both problems.  It identifies the facility that will eventually use these fields.  It also does some minimal exercising of the newly added fields.

Does that seem like a reasonable request?  I'm open to other suggestions if you see other ways to address my concerns,

**Order of Operations**

As software engineers, we all respect that order of operations matters a lot.  This pull request, as it is currently coded, would sit nicely atop the XLS20 pull request once that is merged to develop.  However @nbougalis suggests that we should add this commit _before_ XLS20 is merged.  If we take that approach, then I'm uncomfortable with the pull request as it stands.

If this commit is added before XLS20 is merged, then the pull request presciently adds fields that aren't needed until XLS20 puts them to use.  In effect, if this commit is merged before XLS20 is merged, then I'd like this commit to be a Hooks Only commit.  Let's remove from this commit the fields that XLS20 will introduce.  I'll deal with that problem when I do the rebase of XLS20 to sit atop the 1.9.0 commits.

**Summary**

In summary, assuming we put this commit in before XLS20 is merged, I'd like to see:

1. Adding a test stub file that exercises the features/fields that are being added in anticipation of hooks being merged.
2. Removing the XLS20 fields from the pull request.  We'll deal with adding the XLS20 fields when that rebase happens.

Would it actually be easier for you if we merge this commit after XLS20 is merged?  I'm not intending to make extra work for you (other than adding the unit tests).  Thoughts?  Thanks.

File: src/ripple/protocol/SField.h [hunk 1]
```diff
@@ -150,7 +153,8 @@ class SField
         const char* fn,
         int meta = sMD_Default,
         IsSigning signing = IsSigning::yes);
-    explicit SField(private_access_tag_t, int fc);
+    explicit SField(private_access_tag_t, int fc) = delete;
```
[the remark this answers is not in the document]
Comment (RichardAH) [reply src/ripple/protocol/SField.h:156]: This code is from xls20 you can talk to nikb about why it is the way it is

**Comment by RichardAH** [comment]
I've already had the sfcodes clobbered once, which has delayed Hooks and caused confusion to would-be Hook developers and other people working on ancillary tools.

If Ripple want to encourage external contributions to the code base then they need to become more willing to work collaboratively. This commit is a viable harmonisation commit. If it is not to your liking please reject it and substitute your own. Otherwise continue without harmonising for hooks and demonstrate that Ripple runs the rippled code base for Ripple developers and is not interested in supporting external collaborators.

I really don't see the point arguing about it. Do as you will.

**Comment by RichardAH** [comment]
I have made the changes you have requested.

**Review by scottschurr — APPROVED** [review]
Thanks for making these changes.  I think they will help with the maintenance of the code base, particularly when we look back through history.

I've left a few comments, none of which are critical.  So I'm approving this pull request as it stands.  But it would be nice to see the comments addressed.  So I'm supplying a commit that you may use as an example of how the comments might be addressed.  Or, if you prefer, you could simply cherry-pick the commit on top of your changes.  You'll find the commit here: https://github.com/scottschurr/rippled/commit/94f62f0f0fa988747500f6775c5272da74db881b

Many thanks for your on-going contributions to the XRP Ledger.
